syntax = "proto3";

package clickhouse.grpc;

message QueryInfo {
   string query = 1;
   string query_id = 2;
   map<string, string> settings = 3;
   string database = 4;
   string output_format = 5;
   string user_name = 6;
   string password = 7;
   string quota = 8;
   string session_id = 9;
   bool session_check = 10;
   uint32 session_timeout = 11;
   string input_data = 12;
   bool use_next_input_data = 13;
   bool cancel = 14;
}

message Progress {
   uint64 read_rows = 1;
   uint64 read_bytes = 2;
   uint64 total_rows_to_read = 3;
   uint64 written_rows = 4;
   uint64 written_bytes = 5;
}

message Exception {
   int32 code = 1;
   string name = 2;
   string display_text = 3;
   string stack_trace = 4;
}

message LogEntry {
   uint32 event_time = 1;
   uint32 event_time_microseconds = 2;
   string host_name = 3;
   string query_id = 4;
   uint64 thread_id = 5;
   enum Priority {
      UNKNOWN = 0;
      FATAL = 1;
      CRITICAL = 2;
      ERROR = 3;
      WARNING = 4;
      NOTICE = 5;
      INFORMATION = 6;
      DEBUG = 7;
      TRACE = 8;
   }
   Priority priority = 6;
   string source = 7;
   string text = 8;
}

message Result {
   string output = 1;
   string totals = 2;
   string extremes = 3;
   Progress progress = 4;
   Exception exception = 5;
   bool cancelled = 6;
   repeated LogEntry logs = 7;
}

service ClickHouse {
   rpc ExecuteQuery(QueryInfo) returns (Result) {}
   rpc ExecuteQueryWithStreamingInput(stream QueryInfo) returns (Result) {}
   rpc ExecuteQueryWithStreamingOutput(QueryInfo) returns (stream Result) {}
   rpc ExecuteQueryWithStreaming(stream QueryInfo) returns (stream Result) {}
}
